Crash with no user warning when WebView updates Android 5.x

###### Reproduction Steps

1. Remove Android System Webview updates on Samsung galaxy phone (mostly) Android 5.x
2. Open AnkiDroid
3. Run Play Store updates (to update the Android System WebView)


###### Expected Result

The App *should* reload (part of the webview update) and worst case notify the user the webview is not available


###### Actual Result

The App is reloaded but because it touches the WebView immediately on startup and the WebView is not there, it crashes

Explanation here:
https://bugs.chromium.org/p/chromium/issues/detail?id=506369#c3

Crashes here:
https://play.google.com/apps/publish/?account=5338425030304417978#AndroidMetricsErrorsPlace:p=com.ichi2.anki&appid=4973711737547064258&appVersion=20904300&lastReportedRange=LAST_14_DAYS&installSource=INSTALLED_FROM_ANYWHERE&clusterName=apps/com.ichi2.anki/clusters/d261ebbd&detailsAppVersion=20904300&detailsSpan=14&detailsInstallSource=INSTALLED_FROM_ANYWHERE


Normally I would say ignore it but we will support 5.x for a while longer and this crashes people a lot, 2-3/day so it makes our top-10 crash list.

Also, it should be an easy fix, with the proper fix being to catch this exception bubbling up, test if the webview is available ourselves, and if it is not then throw a dialog with a translated error message telling the user "AnkiDroid relies on the System WebView which is unavailable. This can happen if the system is installing updates. Please try again in a few minutes" and then an app close (since we can't work without a webview anyway)



###### Debug info
Refer to the [support page](https://ankidroid.org/docs/help.html) if you are unsure where to get the "debug info".

###### Research
*Enter an [ x ] character to confirm the points below:*

[  ] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid

[  ] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue

[  ] I have searched for similar existing issues here and on the user forum

@david-allison-1 here is an example:

```
Mar 16, 8:00 AM on app version 20904300
Samsung Galaxy Note3 (hltespr), 3072MB RAM, Android 5.0
Report 1

java.lang.RuntimeException: 
  at android.app.ActivityThread.handleBindApplication (ActivityThread.java:5116)
  at android.app.ActivityThread.access$1600 (ActivityThread.java:177)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1509)
  at android.os.Handler.dispatchMessage (Handler.java:102)
  at android.os.Looper.loop (Looper.java:145)
  at android.app.ActivityThread.main (ActivityThread.java:5942)
  at java.lang.reflect.Method.invoke (Method.java)
  at java.lang.reflect.Method.invoke (Method.java:372)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:1399)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1194)
Caused by: android.util.AndroidRuntimeException: 
  at android.webkit.WebViewFactory.getFactoryClass (WebViewFactory.java:161)
  at android.webkit.WebViewFactory.getProvider (WebViewFactory.java:101)
  at android.webkit.CookieManager.getInstance (CookieManager.java:46)
  at android.webkit.CookieManager.setAcceptFileSchemeCookies (CookieManager.java:309)
  at com.ichi2.anki.AnkiDroidApp.onCreate (AnkiDroidApp.java:220)
  at android.app.Instrumentation.callApplicationOnCreate (Instrumentation.java:1020)
  at android.app.ActivityThread.handleBindApplication (ActivityThread.java:5113)
  at android.app.ActivityThread.access$1600 (ActivityThread.java:177)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1509)
  at android.os.Handler.dispatchMessage (Handler.java:102)
  at android.os.Looper.loop (Looper.java:145)
  at android.app.ActivityThread.main (ActivityThread.java:5942)
  at java.lang.reflect.Method.invoke (Method.java)
  at java.lang.reflect.Method.invoke (Method.java:372)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:1399)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1194)
Caused by: android.content.pm.PackageManager$NameNotFoundException: 
  at android.app.ApplicationPackageManager.getPackageInfo (ApplicationPackageManager.java:137)
  at android.webkit.WebViewFactory.getFactoryClass (WebViewFactory.java:133)
  at android.webkit.WebViewFactory.getProvider (WebViewFactory.java:101)
  at android.webkit.CookieManager.getInstance (CookieManager.java:46)
  at android.webkit.CookieManager.setAcceptFileSchemeCookies (CookieManager.java:309)
  at com.ichi2.anki.AnkiDroidApp.onCreate (AnkiDroidApp.java:220)
  at android.app.Instrumentation.callApplicationOnCreate (Instrumentation.java:1020)
  at android.app.ActivityThread.handleBindApplication (ActivityThread.java:5113)
  at android.app.ActivityThread.access$1600 (ActivityThread.java:177)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:1509)
  at android.os.Handler.dispatchMessage (Handler.java:102)
  at android.os.Looper.loop (Looper.java:145)
  at android.app.ActivityThread.main (ActivityThread.java:5942)
  at java.lang.reflect.Method.invoke (Method.java)
  at java.lang.reflect.Method.invoke (Method.java:372)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:1399)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1194)
```

--- a/AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidApp.java
@@ -177,6 +177,10 @@ public class AnkiDroidApp extends MultiDexApplication {
     /** Our ACRA configurations, initialized during onCreate() */
     private CoreConfigurationBuilder acraCoreConfigBuilder;
 
+    /** An exception if the WebView subsystem fails to load */
+    @Nullable
+    private Throwable mWebViewError;
+
 
     @NonNull
     public static InputStream getResourceAsStream(@NonNull String name) {
@@ -278,7 +282,17 @@ public class AnkiDroidApp extends MultiDexApplication {
         NotificationChannels.setup(getApplicationContext());
 
         // Configure WebView to allow file scheme pages to access cookies.
-        CookieManager.setAcceptFileSchemeCookies(true);
+        try {
+            CookieManager.setAcceptFileSchemeCookies(true);
+        } catch (Throwable e) {
+            // 5794: Errors occur if the WebView fails to load
+            // android.webkit.WebViewFactory.MissingWebViewPackageException.MissingWebViewPackageException
+            // Error may be excessive, but I expect a UnsatisfiedLinkError to be possible here.
+            this.mWebViewError = e;
+            sendExceptionReport(e, "setAcceptFileSchemeCookies");
+            Timber.e(e, "setAcceptFileSchemeCookies");
+            return;
+        }
 
         // Prepare Cookies to be synchronized between RAM and permanent storage.
         CompatHelper.getCompat().prepareWebViewCookies(this.getApplicationContext());
@@ -581,6 +595,22 @@ public class AnkiDroidApp extends MultiDexApplication {
     private static boolean isCurrentLanguage(String l) {
         String pref = getSharedPrefs(sInstance).getString(Preferences.LANGUAGE, "");
         return pref.equals(l) || "".equals(pref) && Locale.getDefault().getLanguage().equals(l);
+    }
+
+
+    public static boolean webViewFailedToLoad() {
+        return getInstance().mWebViewError != null;
+    }
+
+
+    @Nullable
+    public static String getWebViewErrorMessage() {
+        Throwable error = getInstance().mWebViewError;
+        if (error == null) {
+            Timber.w("getWebViewExceptionMessage called without webViewFailedToLoad check");
+            return null;
+        }
+        return error.getLocalizedMessage();
     }
 
     /**

--- a/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java
@@ -81,7 +81,6 @@ import android.widget.Toast;
 import com.afollestad.materialdialogs.MaterialDialog;
 import com.getbase.floatingactionbutton.FloatingActionButton;
 import com.getbase.floatingactionbutton.FloatingActionsMenu;
-import com.ichi2.anim.ActivityTransitionAnimation;
 import com.ichi2.anki.CollectionHelper.CollectionIntegrityStorageCheck;
 import com.ichi2.anki.StudyOptionsFragment.StudyOptionsListener;
 import com.ichi2.anki.analytics.UsageAnalytics;
@@ -573,6 +572,16 @@ public class DeckPicker extends NavigationDrawerActivity implements
      * @return whether or not we were successful
      */
     private boolean firstCollectionOpen() {
+        if (AnkiDroidApp.webViewFailedToLoad()) {
+            new MaterialDialog.Builder(this)
+                    .title(R.string.ankidroid_init_failed_webview_title)
+                    .content(getString(R.string.ankidroid_init_failed_webview, AnkiDroidApp.getWebViewErrorMessage()))
+                    .positiveText(R.string.close)
+                    .onPositive((d, w) -> exit())
+                    .cancelable(false)
+                    .show();
+            return false;
+        }
         if (Permissions.hasStorageAccessPermission(this)) {
             Timber.i("User has permissions to access collection");
             // Show error dialog if collection could not be opened

--- a/AnkiDroid/src/main/res/values/02-strings.xml
+++ b/AnkiDroid/src/main/res/values/02-strings.xml
@@ -330,4 +330,8 @@
     <!-- Deck Picker -->
     <string name="deck_picker_failed_deck_load">Failed to load deck ‘%s’</string>
     <string name="search_decks">Search decks</string>
+
+
+    <string name="ankidroid_init_failed_webview_title">Fatal Error</string>
+    <string name="ankidroid_init_failed_webview">AnkiDroid relies on the System WebView which is unavailable. This can happen if the system is installing updates. Please try again in a few minutes.\n\n%s</string>
 </resources>

